src/x_system/x_remove.1
.TH X_REMOVE 1 "Feb 27 2014" "GMT 4.5.13 (SVN)" "Generic Mapping Tools"
.SH NAME
\fBx_remove\fP \- Remove COEs for Selected Cruises from the Data Base
.SH SYNOPSIS
\fBx_remove\fP \fIremovefile\fP [ \fB\-X\fP\fIxbase\fP ] [ \fB\-L\fP\fIlegbase\fP ] [ \fB\-V\fP ]
.SH DESCRIPTION
\fBx_remove\fP will read a list of 'bad' cruises, read the \fBXSYSTEM\fP data base files, and write out new
\fBXSYSTEM\fP files that do not include any COEs generated by the bad cruises. This is useful if certain
cruises turn out to have systematic errors (e.g., the topography values were inadvertently given as
fathoms instead of meters), and we would like to re-run \fBx_over/x_update\fP after the
systematic errors have been rectified. The new file names will consist of the old file names with
the suffix _new appended.
.TP
\fB\-X\fP
Indicates alternate xbase file [Default is \fIxx_base.b\fP].
.TP
\fB\-L\fP
Indicates alternate legbase file [Default is \fIxx_legs.b\fP].
.TP
\fB\-V\fP
Selects verbose mode, which will send progress reports to stderr [Default runs "silently"].
.SH SEE ALSO
.IR GMT (1),
.IR x_system (1)
.SH REFERENCES
Wessel, P. XOVER: A Cross-over Error Detector for Track Data,
\fIComputers & Geosciences, 15\fP, 333-346.


ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 12925 2014-02-20 22:02:24Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
201x-yy-zz	pwessel
	GMT 5.1.1 finalized and released.
-----------------------------------------------------------------------------------------
2014-02-20	pwessel
	* grdcontour.c:		The -Zscale/offset option was applied incorrectly due to a
				change in the underlying scaling function.
	* grdvolume.c:		---"---
2014-02-10	pwessel
	*gmt2kml.c:		Make sure the default style IDs are unique for each process.

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.
2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		We consider the change in SDIST, PDIST to return degrees instead
				of km to be a bug as it makes it difficult to handle geodesics
				for which degrees make no sense.  Thus we have consolidated all
				the ?DIST operators to give km if geographic data are involved.
				Example 23 has been adjusted accordingly, as has several test
				scripts.  Please check your own scripts for SDIST.

2014-01-09	pwessel
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
2014-01-09	remko
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
	* grdmath.c:		LDIST and PDIST will again produce km when using geodesics.

2014-01-01	pwessel
	* x2sys/x2sys.c:	Mishandled the assignment of segment number for each record.
	* gmt_support.c:	Allow GMT_crossover to deal with global wrapped data.

2013-12-31	pwessel
	* gmt2kml.c:		Now using GMT_Put_Record instead of old stdout printing.
				This addresses issue #473.

2013-12-10	pwessel
	* grdvector.c:		Vector direction did not adjust when negative scales were
				used in -JX or -Jx.
	* psxy[z].c		---"---

2013-12-08	remko
	* psimage.c:		Now properly plots transparent images and makes -Gt work.

2013-11-26	pwessel
	* gmt_io.c:		GMT would crash on a blank line with no newline at end at all.
	* pstext.c:		pstext did not process GMT4-style input properly.

2013-11-26	jluis
	* potential/gravfft.c:	Let -E options default to 3.
				Let users select the -N+l (leave trend alone). That is, do not overwrite with default mode.
				Let the default detrend be remove MID point.
				Explicitly set -N+a in gravfft command to avoid failures due to coming changes in gravfft.
2013-11-25	pwessel
	* grdlandmask.c:	Failed to fill partial bins with no coastlines when the region
				crossed Greenwich [issue #451].

2013-11-21	pwessel
	* pscontour.c:		Let pscontour output contours to file [issue #448].
	* pscoast.c:		Prevent users from not giving proper projection [issue #449].

2013-11-19	jluis
	* mgd77/mgd77magref.c:	Needed to reset the t_col variable when constant altitude [issue #446].

2013-11-19	pwessel
	* gmt_sph.c:		Fix invalid read in ssrfpack_grid due to par array being of length 2 rather than 3.

2013-11-18	pwessel
	* gmt_grdio.c:		Place the old auto-detect of 4-byte binary float vs int under compatibility mode for backwardness.

2013-11-16	pwessel
	* grdvector.c:		Address shortcomings of geovectors when wrapped and clipped.

2013-11-15	pwessel
	* geovectors.c:		Fix geovector when shrinking is selected.

2013-11-13	pwessel
	* gmt_api.c:		Fix API issues pointed out by Ian.

2013-11-13	fwobbe
	* GMT.app:		Inherit user shell in GMT.app.

2013-11-12	pwessel
	* gmt_init.c:		grdmath could corrupt the gmt.history file [issue # 438].
 	
2013-11-10	pwessel
	* gmt_support.c:	Fix symbol-size issue in psxy -Sc [issue #435].
	
2013-11-09	pwessel
	* grdvector.c:		Did not convert direction to azimuths for geo vectors.

2013-11-08	pwessel
	* grdvector.c:		Fix grdvector -I problem [issue #434].
	* ps2raster.c:		Under Mavericks, strcpy would exit if the to and from pointers
-----------------------------------------------------------------------------------------
2013-11-05	pwessel
	GMT 5.1.0 finalized and released.
-----------------------------------------------------------------------------------------

2013-11-01	remko
	* example_10.sh:	Changed input to number of languages; multi-color columns

2013-10-31	pwessel
	* gmt_dcw.c:		Added option to specify code as =<continent> which will
				automatically be expanded to all the countries in that
				continent.  E.g., -F=AF+gred will paint Africa read.
				Fixed memory leaks.

2013-10-23	pwessel
	* gmtinfo.c:		The new name of minmax.c, the latter will be honored
				when GMT compatibility mode is 4.
	* grdmath.c:		Fixed porting bugs for CURV and D2DY2; added test ops.sh

2013-10-22	pwessel
	* minmax.c:		Added -D to adjust region found with -I
	* makecpt.c:		Better reporting of gaps or overlaps in CPTs

2013-10-05	pwessel
	* greenspline.c:	Needed -f in order to select -f0T input, plus it made assumptions
				about lon,lat despite not being selected.

2013-10-04	pwessel
	* gmt.history:		This replaces the old .gmtcommands file name.

2013-09-28	remko
	* gmt_init.c:		Moved origin of JD by one day (24 Nov instead of 25 Nov).
	* ConfigUserTemplate.cmake:	Added option GMT_DOCS_DEPEND_ON_GMT

2013-09-27	pwessel
	* share/cpt:		Added 10 new CPTs, mostly for bathymetry and topography.

2013-09-14	pwessel

src/x_system/x_remove.c
/*	$Id: x_remove.c 9545 2011-07-27 19:31:54Z pwessel $
 *
 * XREMOVE will read a list of bad legs from a file, and then remove all
 * trace of these files from the x_system data base files. New x_system files
 * will be created with the suffix _new appended to the file names.
 * All COE generated by the bad legs will be skipped when writing out the
 * new files, and the mean/stdev statistics will be recalculated
 *
 * Author:	Paul Wessel
 * Date:	25-JAN-1988
 * Last rev:	06-MAR-2000	POSIX
 *
 */
 
#include "gmt.h"
#include "x_system.h"

#define MAXLEGS 10000
#define MAXBADLEGS 1000
#define MAX_X 10000

struct XOVERS crossover[MAX_X];

struct LEG leg[MAXLEGS];

size_t legsize = sizeof (struct LEG);
GMT_LONG nbadlegs = 0, nlegs = 0;

char badlegs[MAXBADLEGS][10];

GMT_LONG findleg (char *name);
GMT_LONG get_id (char *name);

int main (int argc, char **argv)
{
	FILE *fpb = NULL, *fpl = NULL, *fpr = NULL, *fpl2, *fpb2;
	GMT_LONG i, j, n_x, error = FALSE, id1, id2, internal, ok, nrecs = 1, verbose = FALSE;
	char header[BUFSIZ], lega[10], legb[10], line[BUFSIZ];
	double mean;

	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-') {
			switch (argv[i][1]) {
				case 'X':
					fpb = fopen (&argv[i][2], "rb");
					break;
				case 'L':
					fpl = fopen (&argv[i][2], "rb");
					break;
				case 'V':
					verbose = TRUE;
					break;
				default:
					error = TRUE;
					break;
			}
		}
		else
			fpr = fopen (argv[i], "r");
	}

	if (fpr == NULL) error = TRUE;
	if (argc == 1 || error) {
		fprintf(stderr, "xremove - Remove selected info from xover database\n\n");
		fprintf(stderr, "usage: xremove removefile [-X<xbasefile> -L<xlegsfile> -V]\n");
		fprintf(stderr, "	removefile is a list of legs to remove\n");
		fprintf(stderr, "	-X to select alternate xx_base.b file\n");
		fprintf(stderr, "	-L to select alternate xx_legs.b file\n");
		fprintf(stderr, "	-V Verbose, report when removing a leg\n");
		exit (EXIT_FAILURE);
	}

	if (fpb == NULL && (fpb = fopen("xx_base.b","rb")) == NULL) {
		fprintf (stderr, "Could not find xx_base.b\n");
		exit (EXIT_FAILURE);
	}
	if (fpl == NULL && (fpl = fopen("xx_legs.b","rb")) == NULL) {
		fprintf (stderr, "Could not find xx_legs.b\n");
		exit (EXIT_FAILURE);
	}

	/* Read all the leg info into memory */

	nlegs = 0;
	while (fread ((void *)(&leg[nlegs]), legsize, (size_t)1, fpl) == (size_t)1) {
		for (j = 0; j < 3; j++) {
			leg[nlegs].n_gmtint[j] = leg[nlegs].n_gmtext[j] = 0;
			leg[nlegs].mean_gmtint[j] = leg[nlegs].mean_gmtext[j] = 0.0;
			leg[nlegs].st_dev_gmtint[j] = leg[nlegs].st_dev_gmtext[j] = 0.0;
			leg[nlegs].dc_shift_gmt[j] = leg[nlegs].drift_rate_gmt[j] = 0.0;
		}
		leg[nlegs].n_x_int = leg[nlegs].n_x_ext = 0;
		leg[nlegs].next_leg = 0;
		nlegs++;
		if (nlegs == MAXLEGS) {
			fprintf (stderr, "xremove: nlegs > MAXLEGS, recompile!\n");
			exit (EXIT_FAILURE);
		}
	}
	fclose (fpl);

	while (fgets (line, BUFSIZ, fpr)) {
		sscanf (line, "%s", badlegs[nbadlegs]);
		nbadlegs++;
		if (nbadlegs == MAXBADLEGS) {
			fprintf (stderr, "xremove: nbadlegs > MAXBADLEGS, recompile!\n");
			exit (EXIT_FAILURE);
		}
	}
	fclose (fpr);

	if ((fpb2 = fopen("xx_base.b_new", "wb")) == NULL) {
		fprintf (stderr, "Could not create xx_base.b_new\n");
		exit (EXIT_FAILURE);
	}
	if ((fpl2 = fopen("xx_legs.b_new","wb")) == NULL) {
		fprintf (stderr, "Could not create xx_legs.b_new\n");
		exit (EXIT_FAILURE);
	}

	if (fread((void *)header, (size_t)REC_SIZE, (size_t)1, fpb) != (size_t)1) {
		fprintf (stderr, "xremove: Read error on xx_base.b\n");
		exit (EXIT_FAILURE);
	}

	if (fwrite ((void *)header, (size_t)REC_SIZE, (size_t)1, fpb2) != (size_t)1) {
		fprintf (stderr, "xremove: Write error on xx_base.b\n");
		exit (EXIT_FAILURE);
	}
	ok = fread ((void *)header, (size_t)REC_SIZE, (size_t)1, fpb);
	while (ok) {
		sscanf(header, "%s %s %" GMT_LL "d",lega, legb, &n_x);
		if (findleg (lega) || findleg (legb)) {
			if (verbose) fprintf (stderr, "xremove: Skipping %s - %s\n", lega, legb);
			fseek (fpb, (long int)(REC_SIZE*n_x), SEEK_CUR);
		}
		else {
			internal = !strcmp(lega, legb);
			id1 = get_id (lega);
			id2 = (internal) ? id1 : get_id (legb);
			if (n_x > MAX_X) {
				fprintf (stderr, "xremove: nx (= %ld) > MAX_X, recompile!\n", n_x);
				exit (EXIT_FAILURE);
			}
			if (fread ((void *)crossover, (size_t)REC_SIZE, (size_t)n_x, fpb) != (size_t)n_x) {
          			fprintf (stderr, "xremove: Read error on xx_base.b\n");
          			exit (EXIT_FAILURE);
          		}
          		if (fwrite ((void *)header, (size_t)REC_SIZE, (size_t)1, fpb2) != (size_t)1) {
          			fprintf (stderr, "xremove: Write error on xx_base.b\n");
          			exit (EXIT_FAILURE);
          		}
          		for (i = 0; i < n_x; i++) {
          			if (internal) {
          				for (j = 0; j < 3; j++) {
          					if (crossover[i].x_val[j] == NODATA) continue;
          					leg[id1].mean_gmtint[j] += crossover[i].x_val[j];
          					leg[id1].st_dev_gmtint[j] += crossover[i].x_val[j]*crossover[i].x_val[j];
          					leg[id1].n_gmtint[j]++;
          				}
          			}
          			else {
          				for (j = 0; j < 3; j++) {
          					if (crossover[i].x_val[j] == NODATA) continue;
          					leg[id1].mean_gmtext[j] += crossover[i].x_val[j];
          					leg[id1].st_dev_gmtext[j] += crossover[i].x_val[j]*crossover[i].x_val[j];
          					leg[id1].n_gmtext[j]++;
          					leg[id2].mean_gmtext[j] -= crossover[i].x_val[j];
          					leg[id2].st_dev_gmtext[j] += crossover[i].x_val[j]*crossover[i].x_val[j];
          					leg[id2].n_gmtext[j]++;
          				}
          			}
          		}
           		if (fwrite ((void *)crossover, (size_t)REC_SIZE, (size_t)n_x, fpb2) != (size_t)n_x) {
           			fprintf (stderr, "xremove: Write error on xx_base.b\n");
           			exit (EXIT_FAILURE);
           		}
         		if (internal)
          			leg[id1].n_x_int = (int)n_x;
          		else {
          			leg[id1].n_x_ext += (int)n_x;
          			leg[id2].n_x_ext += (int)n_x;
          		}
          		nrecs += n_x + 1;
          	}
		ok = fread ((void*)header, (size_t)REC_SIZE, (size_t)1, fpb);
	}
	fclose (fpb);
	sprintf (header, "%10ld xx_base.b header", nrecs);
	fseek (fpb2, 0L, SEEK_SET);
	if (fwrite ((void *)header, (size_t)REC_SIZE, (size_t)1, fpb2) != (size_t)1) {
		fprintf (stderr, "xremove: Write error on xx_base.b\n");
		exit (EXIT_FAILURE);
	}
	fclose (fpb2);

	if (verbose) fprintf (stderr, "xremove: New xx_legs.b-file created successfully\n");

	for (i = 0; i < nlegs; i++) {
		if (findleg (leg[i].name)) continue;

		for (j = 0; j < 3; j++) {
			/* Compute statistics for internal crossovers */

			mean = (leg[i].n_gmtint[j]) ? leg[i].mean_gmtint[j] / leg[i].n_gmtint[j] : 0.0;
			leg[i].st_dev_gmtint[j] = (leg[i].n_gmtint[j] > 1) ? 
				sqrt((leg[i].st_dev_gmtint[j]-mean*leg[i].mean_gmtint[j])/(leg[i].n_gmtint[j]-1)) : 0.0;
			leg[i].mean_gmtint[j] = mean;

			/* Same for external crossovers */

			mean = (leg[i].n_gmtext[j]) ? leg[i].mean_gmtext[j] / leg[i].n_gmtext[j] : 0.0;
			leg[i].st_dev_gmtext[j] = (leg[i].n_gmtext[j] > 1) ?
				sqrt((leg[i].st_dev_gmtext[j]-mean*leg[i].mean_gmtext[j])/(leg[i].n_gmtext[j]-1)) : 0.0;
			leg[i].mean_gmtext[j] = mean;
		}
	}
	if (fwrite ((void *)leg, legsize, (size_t)nlegs, fpl2) != (size_t)nlegs) {
		fprintf (stderr, "xremove: Write error for xx_legs.b file\n");
		exit (EXIT_FAILURE);
	}
	fclose (fpl2);
	if (verbose) fprintf (stderr, "xremove: New xx_legs.b-file created successfully\n");
	exit (EXIT_SUCCESS);
}
          
GMT_LONG findleg (char *name)
{
	GMT_LONG left, right, mid, cmp;

	left = 0;
	right = nbadlegs-1;
	while (left <= right) {
		mid = (left + right)/2;
		cmp = strcmp(name, badlegs[mid]);
		if (cmp < 0)
			right = mid-1;
		else if (cmp > 0)
			left = mid+1;
		else
			return (1);
	}
	return (0);
}

GMT_LONG get_id (char *name)
{
	GMT_LONG left, right, mid, cmp;

	left = 0;
	right = nlegs-1;
	while (left <= right) {
		mid = (left + right)/2;
		cmp = strcmp(name, leg[mid].name);
		if (cmp < 0)
			right = mid-1;
		else if (cmp > 0)
			left = mid+1;
		else
			return (mid);
	}
	return (-1);
}
